core/input/devices.js

```diff
@@ -50,59 +45,220 @@
+            }
+
+            var last;
+            if (this._keyDownList.length === 0) {
+                last = null;
             } else {
-                // Allow the event to bubble and become a keyPress event which
-                // will have the character code translated
-                return true;
+                last = this._keyDownList[this._keyDownList.length-1];
+            }
+
+            // insert a new entry if last seen key was different.
+            if (!last || code === 'Unidentified' || last.code !== code) {
+                last = {code: code, keysyms: {}};
+                this._keyDownList.push(last);
+            }
+
+            // Wait for keypress?
+            if (!keysym) {
+                return;
+            }
+
+            // make sure last event contains this keysym (a single "logical" keyevent
+            // can cause multiple key events to be sent to the VNC server)
+            last.keysyms[keysym] = keysym;
+            last.ignoreKeyPress = true;
+
+            // undo modifiers
+            if (escape) {
+                for (var i = 0; i < escape.length; ++i) {
+                    this._sendKeyEvent(escape[i], 'Unidentified', false);
+                }
+            }
+
+            // send the character event
+            this._sendKeyEvent(keysym, code, true);
+
+            // redo modifiers
+            if (escape) {
+                for (i = 0; i < escape.length; ++i) {
+                    this._sendKeyEvent(escape[i], 'Unidentified', true);
+                }
             }
         },
 
+        // Legacy event for browsers without code/key
         _handleKeyPress: function (e) {
             if (!this._focused) { return true; }
 
-            if (this._handler.keypress(e)) {
-                // Suppress bubbling/default actions
-                Util.stopEvent(e);
-                return false;
+            Util.stopEvent(e);
+
+            var code = this._getKeyCode(e);
+            var keysym = KeyboardUtil.getKeysym(e);
+
+            // if a char modifier is pressed, get the keys it consists
+            // of (on Windows, AltGr is equivalent to Ctrl+Alt)
+            var active = this._modifierState.activeCharModifier();
+
+            // If we have a char modifier down, and we're able to
+            // determine a keysym reliably then (a) we know to treat
+            // the modifier as a char modifier, and (b) we'll have to
+            // "escape" the modifier to undo the modifier when sending
+            // the char.
+            if (active && keysym) {
+                var isCharModifier = false;
+                for (var i  = 0; i < active.length; ++i) {
+                    if (active[i] === keysym) {
+                        isCharModifier = true;
+                    }
+                }
+                if (!isCharModifier) {
+                    var escape = this._modifierState.activeCharModifier();
+                }
+            }
+
+            var last;
```

Comment:
please comment this part as well, explain the list (it's not as clear as when things were broken up separately)